lti_consumer/lti_consumer.py

```diff
@@ -1,982 +1,292 @@
 
     Returns:
-        unicode: anon_id if match found
+        (float, str):  (score, [optional]comment) if parsing is successful
 
     Raises:
-        LtiError if suffix cannot be parsed or is not in its expected form
+        LtiError: if verification fails
     """
-    if suffix:
-        match_obj = RESULT_SERVICE_SUFFIX_PARSER.match(suffix)
-        if match_obj:
-            return match_obj.group('anon_id')
-    # fall-through handles all error cases
-    msg = _("No valid user id found in endpoint URL")
-    log.info("[LTI]: %s", msg)
-    raise LtiError(msg)
-
+    try:
+        json_obj = json.loads(json_str)
+    except (ValueError, TypeError):
+        msg = "Supplied JSON string in request body could not be decoded: {}".format(json_str)
+        log.error("[LTI] %s", msg)
+        raise LtiError(msg)
+
+    # The JSON object must be a dict. If a non-empty list is passed in,
+    # use the first element, but only if it is a dict
+    if isinstance(json_obj, list) and len(json_obj) >= 1:
+        json_obj = json_obj[0]
+
+    if not isinstance(json_obj, dict):
+        msg = ("Supplied JSON string is a list that does not contain an object as the first element. {}"
+               .format(json_str))
+        log.error("[LTI] %s", msg)
+        raise LtiError(msg)
+
+    # '@type' must be "Result"
+    result_type = json_obj.get("@type")
+    if result_type != "Result":
+        msg = "JSON object does not contain correct @type attribute (should be 'Result', is z{})".format(result_type)
+        log.error("[LTI] %s", msg)
+        raise LtiError(msg)
+
+    # '@context' must be present as a key
+    if '@context' not in json_obj:
+        msg = "JSON object does not contain required key @context"
+        log.error("[LTI] %s", msg)
+        raise LtiError(msg)
+
+    # Return None if the resultScore key is missing, this condition
+    # will be handled by the upstream caller of this function
+    if "resultScore" not in json_obj:
+        score = None
+    else:
+        # if present, 'resultScore' must be a number between 0 and 1 inclusive
+        try:
+            score = float(json_obj.get('resultScore', "unconvertable"))  # Check if float is present and the right type
+            if not 0.0 <= score <= 1.0:
+                msg = 'score value outside the permitted range of 0.0-1.0.'
+                log.error("[LTI] %s", msg)
+                raise LtiError(msg)
+        except (TypeError, ValueError) as err:
+            msg = "Could not convert resultScore to float: {}".format(str(err))
+            log.error("[LTI] %s", msg)
+            raise LtiError(msg)
 
-LaunchTargetOption = namedtuple('LaunchTargetOption', ['display_name', 'value'])
+    return score, json_obj.get('comment', "")
 
 
-class LaunchTarget(object):  # pylint: disable=bad-option-value, useless-object-inheritance
-    """
-    Constants for launch_target field options
+class LtiConsumer(LtiConsumerBase):
```

Comment:
Don't use class inheritance like this. 
Instead make the LTI code independent from the XBlock structure and instance the class in the block when calling the launch parameters.

The main motivation for this comment is that this will make a things more explicit and force a boundary between the framework/storage independent LTI 1.1 consumer and the XBlock structure.

In the 1.3 implementation, I explicitly instance the LTI class when using it, and pass everything as parameters, as opposed to have an adapter class in the middle: https://github.com/edx/xblock-lti-consumer/blob/opencraftbd/lti_1p3/lti_consumer/lti_consumer.py#L807.

Do you think it's possible to do this here too? Do you see any issue with that approach?
@xitij2000 Feel free to pitch in here too.

Reply:
I think I recommended inheritance during our call but in retrospect seeing how it's done with lti1.3 that approach is definitely better. 
@pcockwell Sorry I think you were right to suggest composition here.